presto-thrift-connector/src/main/java/com/facebook/presto/connector/thrift/ThriftFilterPushdown.java

```diff
@@ -0,0 +1,195 @@
+
+/**
+ * Runs during both logical and physical phases of connector-aided plan optimization.
+ * In most cases filter pushdown will occur during logical phase. However, in cases
+ * when new filter is added between logical and physical phases, e.g. a filter on a join
+ * key from one side of a join is added to the other side, the new filter will get
+ * merged with the one already pushed down.
+ */
+public class ThriftFilterPushdown
+        implements ConnectorPlanOptimizer
+{
+    private final ThriftMetadata metadata;
+    protected final RowExpressionService rowExpressionService;
+
+    public ThriftFilterPushdown(ThriftMetadata metadata, RowExpressionService rowExpressionService)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.rowExpressionService = requireNonNull(rowExpressionService, "rowExpressionService is null");
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode maxSubplan, ConnectorSession session, VariableAllocator variableAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        return rewriteWith(new Rewriter(session, idAllocator, metadata), maxSubplan);
+    }
+
+    public static class Rewriter
+            extends ConnectorPlanRewriter<Void>
+    {
+        private final ConnectorSession session;
+        private final PlanNodeIdAllocator idAllocator;
+        private final ConnectorMetadata metadata;
+
+        public Rewriter(ConnectorSession session, PlanNodeIdAllocator idAllocator, ConnectorMetadata metadata)
+        {
+            this.session = requireNonNull(session, "session is null");
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.metadata = requireNonNull(metadata, "metadata is null");
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode project, RewriteContext<Void> context)
+        {
+            // Thrift connector in Prestissimo must have no other PlanNode be present
+            // between IndexJoinNode and IndexSourceNode.
+            for (Map.Entry<VariableReferenceExpression, RowExpression> entry : project.getAssignments().entrySet()) {
+                if (!entry.getKey().equals(entry.getValue())) {
+                    return visitPlan(project, context);
+                }
+            }
+            // If all assignments are identical, remove the current ProjectNode.
+            return visitPlan(project.getSource(), context);
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode filter, RewriteContext<Void> context)
+        {
+            if (!(filter.getSource() instanceof IndexSourceNode)) {
+                return visitPlan(filter, context);
+            }
+
+            IndexSourceNode indexSource = (IndexSourceNode) filter.getSource();
+            if (!(indexSource.getTableHandle().getConnectorHandle() instanceof ThriftTableHandle)
+                    || indexSource.getTableHandle().getLayoutHandle() == null
```

Comment:
in which case getLayoutHandle is null or not ThriftTableLayoutHandle? Should we have checks on these?